File: mstparser/FeatureVector.java
package mstparser;

import gnu.trove.*;
import java.util.*;

public class FeatureVector {

    public int index;
    public double value;
    public FeatureVector next;
	

    public FeatureVector(int[] keys) {
	FeatureVector result = new FeatureVector();
	for (int i=0; i<keys.length-1; i++)
	    result = new FeatureVector(keys[i],1.0,result);

	index = keys[keys.length-1];
	value = 1.0;
	next = result;
    }

    public FeatureVector() {
	index = -1;
	value = -1.0;
	next = null;
    }
    
    public FeatureVector(int i, double v, FeatureVector n) {
	index = i;
	value = v;
	next = n;
    }

    public int[] keys() {
	TIntArrayList keys = new TIntArrayList();
	for(FeatureVector curr = this; curr.next != null; curr = curr.next) {
	    if(curr.index < 0)
		continue;
	    keys.add(curr.index);
	}
	return keys.toNativeArray();

    }

    public FeatureVector cat(FeatureVector fv2) {
	FeatureVector result = new FeatureVector();
	for(FeatureVector curr = this; curr.next != null; curr = curr.next) {
	    if(curr.index < 0)
		continue;
	    result = new FeatureVector(curr.index,curr.value,result);
	}
	for(FeatureVector curr = fv2; curr.next != null; curr = curr.next) {
	    if(curr.index < 0)
		continue;
	    result = new FeatureVector(curr.index,curr.value,result);
	}
	return result;
		
    }

    // fv1 - fv2
    public FeatureVector getDistVector(FeatureVector fv2) {
	FeatureVector result = new FeatureVector();
	for(FeatureVector curr = this; curr.next != null; curr = curr.next) {
	    if(curr.index < 0)
		continue;
	    result = new FeatureVector(curr.index,curr.value,result);
	}
	for(FeatureVector curr = fv2; curr.next != null; curr = curr.next) {
	    if(curr.index < 0)
		continue;
	    result = new FeatureVector(curr.index,-curr.value,result);
	}
	return result;
    }
	
    public double dotProduct(FeatureVector fv2) {
	double result = 0.0;
	TIntDoubleHashMap hm1 = new TIntDoubleHashMap();
	TIntDoubleHashMap hm2 = new TIntDoubleHashMap();

	for(FeatureVector curr = this; curr.next != null; curr = curr.next) {
	    if(curr.index < 0)
		continue;
	    hm1.put(curr.index,hm1.get(curr.index)+curr.value);
	}
	for(FeatureVector curr = fv2; curr.next != null; curr = curr.next) {
	    if(curr.index < 0)
		continue;
	    hm2.put(curr.index,hm2.get(curr.index)+curr.value);
	}

	int[] keys = hm1.keys();

	for(int i = 0; i < keys.length; i++) {
	    double v1 = hm1.get(keys[i]);
	    double v2 = hm2.get(keys[i]);
	    result += v1*v2;
	}
		
	return result;
		
    }

    public int sum() {
	if(next == null)
	    return index >= 0 ? 1 : 0;
	return (index >= 0 ? 1 : 0) + next.sum();
    }

    public void sort() {
	int[] feats = new int[sum()];
	int j = 0;
	for(FeatureVector curr = this; curr.next != null; curr = curr.next) {
	    if(curr.index < 0)
		continue;
	    feats[j] = curr.index;
	    j++;
	}
	Arrays.sort(feats);
	FeatureVector result = new FeatureVector();
	for(int i = feats.length-1; i >= 0; i--)
	    result = new FeatureVector(feats[i],1.0,result);
	this.index = result.index; this.value = result.value; this.next = result.next;
    }

    public String toString() {
	if(next == null)
	    return ""+index;
	return index + " " + next.toString();
    }

}

File: src/main/scala/mstparser/FeatureVector.scala
package mstparser

import gnu.trove.list.array.TIntArrayList
import gnu.trove.map.hash.TIntDoubleHashMap

object FeatureVector {
  def fromKeys(keys: Array[Int]) = {
    val v = new FeatureVector
    keys.foreach(k => v.add(k, 1.0))
    v
  }
}

class FeatureVector(fv1: FeatureVector, fv2: FeatureVector, negSecond: Boolean) {
  private val features = scala.collection.mutable.Buffer.empty[Feature]
  def this(fv1: FeatureVector, fv2: FeatureVector) = this(fv1, fv2, false)
  def this(fv1: FeatureVector) = this(fv1, null)
  def this() = this(null)

  def cat(that: FeatureVector) = new FeatureVector(this, that)
  def getDistVector(that: FeatureVector) = new FeatureVector(this, that, true)

  def add(i: Int, v: Double) {
    this.features += new Feature(i, v)
  }

  def keys = {
    val keys = new TIntArrayList
    this.addKeysToList(keys)
    keys.toArray
  }

  private def addKeysToList(keys: TIntArrayList) {
    Option(this.fv1).foreach(_.addKeysToList(keys))
    Option(this.fv2).foreach(_.addKeysToList(keys))

    this.features.foreach(f => keys.add(f.index))
  }

  def getScore(parameters: Array[Double]): Double = this.getScore(parameters, false)

  private def getScore(parameters: Array[Double], negate: Boolean): Double = {
    var score = Option(this.fv1).map(_.getScore(parameters, negate)).getOrElse(0.0) +
      Option(this.fv2).map(_.getScore(parameters, negate != this.negSecond)).getOrElse(0.0)

    val mult = if (negate) -1.0 else 1.0
    score + this.features.map(f => mult * parameters(f.index) * f.value).sum
  }

  def update(parameters: Array[Double], total: Array[Double], alpha: Double, update: Double) {
    this.update(parameters, total, alpha, update, false)
  }

  private def update(parameters: Array[Double], total: Array[Double], alpha: Double, update: Double, negate: Boolean) {
    Option(this.fv1).foreach(_.update(parameters, total, alpha, update, negate))
    Option(this.fv2).foreach(_.update(parameters, total, alpha, update, negate != this.negSecond))

    val mult = if (negate) -1.0 else 1.0

    this.features.foreach { f =>
      parameters(f.index) += mult * alpha * f.value
      total(f.index) += mult * alpha * f.value * update
    }
  }

  def dotProduct(that: FeatureVector) = {
    val m1 = new TIntDoubleHashMap(this.features.size)
    this.addFeaturesToMap(m1, false)
    m1.compact()

    val m2 = new TIntDoubleHashMap(that.features.size)
    that.addFeaturesToMap(m2, false)
    m2.compact()

    m1.keys.map(k => m1.get(k) * m2.get(k)).sum
  }

  private def addFeaturesToMap(map: TIntDoubleHashMap, negate: Boolean) {
    Option(this.fv1).foreach(_.addFeaturesToMap(map, negate))
    Option(this.fv2).foreach(_.addFeaturesToMap(map, negate != this.negSecond))

    val mult = if (negate) -1.0 else 1.0

    this.features.foreach { f =>
      if (!map.adjustValue(f.index, mult * f.value))
        map.put(f.index, mult * f.value)
    }
  }

  override def toString = {
    val builder = new StringBuilder
    this.toString(builder)
    builder.toString
  }

  private def toString(builder: StringBuilder) {
    Option(this.fv1).foreach(_.toString(builder))
    Option(this.fv2).foreach(_.toString(builder))

    this.features.foreach(f => builder.append(f.toString).append(' '))
  }
}

